File: lib/env.dart
import 'cons.dart';
import 'datatypes.dart';

class Env {
  static final Map<SymbExpr, Object> _globals = {};

  final Cons _cons;

  Env([this._cons]);

  static void bindGlobal(SymbExpr symb, Object obj) {
    _globals[symb] = obj;
  }

  static void bindStringGlobal(String symb, Object obj) {
    _globals[SymbExpr(symb)] = obj;
  }

  static void resetGlobalEnv() {
    _globals.clear();
  }

  static Object lookupGlobal(SymbExpr sym) {
    var obj = _globals[sym];
    if (obj == null) {
      throw Exception('Symbol binding ${sym} not found in env');
    }
    return obj;
  }

  Object lookup(SymbExpr symb) {
    if (_cons != null) {
      var lookup = _cons.lookup(symb);
      if (lookup != null) {
        return lookup;
      }
    }
    return lookupGlobal(symb);
  }

  Env bind(SymbExpr symb, Object val) {
    return Env(Cons(Cons(symb, val), _cons));
  }

  // TODO base env
}

File: app/src/main/java/gr/cons/nancy/lang/Env.java
package gr.cons.nancy.lang;

import gr.cons.nancy.lang.App.NancyExpr;

import java.util.HashMap;
import java.util.Map;

import static gr.cons.nancy.lang.App.*;

public class Env {

    private final NancyExpr cons;

    public Env() {
        this.cons = NilExpr.NIL;
    }
    public Env(NancyExpr cons) {
        this.cons = cons;
    }

    public static Env of(Cons cons) {
        return new Env(cons);
    }

    public NancyExpr lookup(SymbExpr sym) {
        if (cons != NilExpr.NIL) {
            NancyExpr lookup = ((Cons) cons).lookup(sym);
            if (lookup != null) {
                return lookup;
            }
        }
        throw new IllegalArgumentException("Symbol " + sym + " not found in env");
    }

    public Env bind(SymbExpr sym, NancyExpr val) {
        return of(new Cons(new Cons(sym, val), cons));
    }

    @Override
    public String toString() {
        return "" + cons;
    }

}
